securedrop/tests/migrations/migration_15ac9509fc68.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+
+import random
+import string
+
+from datetime import datetime
+from sqlalchemy import text
+
+from db import db
+from journalist_app import create_app
+
+random.seed('ᕕ( ᐛ )ᕗ')
+
+
+def random_bool():
+    return bool(random.getrandbits(1))
+
+
+def bool_or_none():
+    return random.choice([None, True, False])
+
+
+def random_bytes(min, max, nullable):
+    if nullable and random_bool():
+        return None
+    else:
+        # python2 just wants strings, fix this in python3
+        return random_chars(random.randint(min, max))
+
+
+def random_username():
+    len = random.randint(3, 64)
+    return random_chars(len)
+
+
+def random_chars(len, chars=string.printable):
+    return ''.join([random.choice(chars) for _ in range(len)])
+
+
+def random_datetime(nullable):
+    if nullable and random_bool():
+        return None
+    else:
+        return datetime(
+            year=random.randint(1, 9999),
+            month=random.randint(1, 12),
+            day=random.randint(1, 28),
+            hour=random.randint(0, 23),
+            minute=random.randint(0, 59),
+            second=random.randint(0, 59),
+            microsecond=random.randint(0, 1000),
+        )
+
+
+class UpgradeTester():
+
+    '''This migration has no upgrade because there are no tables in the
+       database prior to running, so there is no data to load or test.
+    '''
+
+    def __init__(self, config):
+        pass
+
+    def load_data(self):
+        pass
+
+    def check_upgrade(self):
+        pass
+
+
+class DowngradeTester():
+
+    def __init__(self, config):
+        self.config = config
+        self.app = create_app(config)
+
+    def load_data(self):
+        with self.app.app_context():
+            for _ in range(200):
+                self.add_journalist()
+
+            for _ in range(200):
+                self.add_source()
+
+            for jid in range(1, 200, 10):
+                for _ in range(random.randint(1, 3)):
+                    self.add_journalist_login_attempt(jid)
+
+            for jid in range(1, 200, 10):
+                for sid in range(1, 200, 10):
+                    self.add_reply(jid, sid)
+
+            for sid in range(1, 200, 10):
+                self.add_source_star(sid)
+
+            for sid in range(1, 200, 8):
+                for _ in range(random.randint(1, 3)):
+                    self.add_submission(sid)
```

Comment:
So in order to test the broken FK submission case in #1189 we could have an `add_orphaned_submissions` static method that adds some submissions, and then deletes their sources. Nice thing: since the cascade delete is enforced at the ORM level and not the database, then the submissions will not be deleted. Also, if we want to defer including this particular case to the Mega Pre-Release Test script I'm cool with that

Reply:
Fixed.